src/feat/pitch-functions.cc

```diff
@@ -835,19 +844,35 @@ void OnlinePitchFeatureImpl::ExtractFrame(
   int32 offset = static_cast<int32>(sample_index -
                                     downsampled_samples_processed_);
 
+  // Treat edge cases first
+  if (sample_index < 0) {
+    // Part of the frame is before the beginning of the signal. This
+    // should only happen if opts_.snip_edges == false, when we are
+    // processing the first few frames of signal. In this case
+    // we pad with zeros.
+    KALDI_ASSERT(opts_.snip_edges == false);
```

Comment:
OK, that sounds fine.
So I guess this is ready to merge.  You're sure this doesn't change results
with snip-edges=true (the default) ?


On Tue, Jun 21, 2016 at 7:30 AM, Blaise Potard <notifications@github.com>
wrote:

> In src/feat/pitch-functions.cc
> <https://github.com/kaldi-asr/kaldi/pull/823#discussion_r67851348>:
>
> > @@ -835,19 +844,35 @@ void OnlinePitchFeatureImpl::ExtractFrame(
> >    int32 offset = static_cast<int32>(sample_index -
> >                                      downsampled_samples_processed_);
> >
> > +  // Treat edge cases first
> > +  if (sample_index < 0) {
> > +    // Part of the frame is before the beginning of the signal. This
> > +    // should only happen if opts_.snip_edges == false, when we are
> > +    // processing the first few frames of signal. In this case
> > +    // we pad with zeros.
> > +    KALDI_ASSERT(opts_.snip_edges == false);
>
> Normally the case where we have a pathological number of samples should be
> handled correctly:
>
>    1. It will first go through the "sample_index < 0" case, which only
>    fills the beginning of the window with zeros, then call the Extract_frame
>    function again with a sample_index of 0 and a reduced window.
>    2. On the second call it will go through the offset +
>    full_frame_length > downsampled_wave_part.Dim() branch, which will fill the
>    end of the window with 0, then call the function again on a sub-window
>    3. On the 3rd call It will go in the general case, filling up what's
>    left of the window with only valid data.
>
> The window on the 3rd call should have exactly the same length as the
> number of samples in the input; although with an insufficient number of
> samples I am not sure what the tracker will output.
>
> I can add some test for pathological cases if you think that could be
> useful!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/823/files/6a9b8a25c9129b57b84560cd60d2c9b0bd9e806b#r67851348>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADJVu-T-okpcW5SGE3c9JRgu5KwjOdLnks5qN8tagaJpZM4IqgEB>
> .
>